File: backend/app/schemas/chatbot.py
"""Pydantic schemas for the AI Reporting Chatbot."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ConversationMessage(BaseModel):
    """A single message in the conversation history."""
    role: str = Field(..., pattern="^(user|bot)$")
    content: str


class ChatbotParseRequest(BaseModel):
    """Request body for the chatbot parse endpoint."""
    message: str = Field(min_length=1, max_length=2000)
    conversation: list[ConversationMessage] = []
    extracted_fields: dict = {}


class ExtractedFields(BaseModel):
    """Fields extracted from the conversation so far."""
    title: str | None = None
    description: str | None = None
    category_id: str | None = None
    category_name: str | None = None
    zone_id: str | None = None
    zone_name: str | None = None


class ChatbotParseResponse(BaseModel):
    """Response from the chatbot parse endpoint."""
    bot_message: str
    fields: ExtractedFields
    status: str = Field(..., description="conversing | ready")
    suggestions: list[str] = []

File: frontend/src/lib/chatbot.ts
/**
 * Chatbot API types and helper functions.
 */
import { api } from "./api";

// ── Types ────────────────────────────────────────────────────────────────────

export interface ConversationMessage {
  role: "user" | "bot";
  content: string;
}

export interface ExtractedFields {
  title: string | null;
  description: string | null;
  category_id: string | null;
  category_name: string | null;
  zone_id: string | null;
  zone_name: string | null;
}

export interface ChatbotParseResponse {
  bot_message: string;
  fields: ExtractedFields;
  status: "conversing" | "ready";
  suggestions: string[];
}

// ── API Functions ────────────────────────────────────────────────────────────

export async function parseChatbotMessage(
  message: string,
  conversation: ConversationMessage[],
  extractedFields: Record<string, unknown>
): Promise<ChatbotParseResponse> {
  return api.post<ChatbotParseResponse>("/chatbot/parse", {
    message,
    conversation,
    extracted_fields: extractedFields,
  });
}
